### PR TITLE
[main][feat] swap helper

### PR DESCRIPTION
**Pull request by wow-alpaca**
## Description
- add SwapHelper contract

## Why?
- to store swap calldata for modification

### CODE REVIEW

File: solidity/tests/swap-helper/SwapHelper.t.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      })
+    );
+
+    // get swap calldata with same amountIn and to should be the same
+    bytes memory _retrievedCalldata = swapHelper.getSwapCalldata(_token0, _token1, _amountIn, _to);
+    assertEq(keccak256(_retrievedCalldata), keccak256(_calldata));
+  }
+
+  function testRevert_SetSwapInfo_WhenCallerIsNotOwner() public {
+    vm.expectRevert();
```
Comment (69alpacas) [inline solidity/tests/swap-helper/SwapHelper.t.sol:147]: add revert reason

File: solidity/contracts/swap-helper/SwapHelper.sol [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  function _replace(
+    bytes memory _data,
+    address _addr,
+    uint256 _offset
+  ) internal pure {
+    assembly {
+      let dataPointer := add(_data, 32)
+      let skipFuncSig := add(4, dataPointer)
+      mstore(add(_offset, skipFuncSig), _addr)
+    }
```
Comment (69alpacas) [inline solidity/contracts/swap-helper/SwapHelper.sol:54]: add comment what each line does and why

File: solidity/contracts/swap-helper/SwapHelper.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+import { ISwapHelper } from "../interfaces/ISwapHelper.sol";
+
+contract SwapHelper is ISwapHelper, Ownable {
+  // source => desination => SwapInfo
+  mapping(address => mapping(address => SwapInfo)) public swapInfos;
+
+  constructor() {}
```
Comment (69alpacas) [inline solidity/contracts/swap-helper/SwapHelper.sol:12]:
```suggestion
```

File: solidity/tests/swap-helper/SwapHelper.t.sol [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      toOffset: 64 + 4
+    });
+
+    swapHelper.setSwapInfo(_token0, _token1, _swapInfo);
+
+    // get swap calldata with modified amountIn and to
+
+    uint256 _newAmountIn = 200e18;
+    address _newTo = RECIPIENT_2;
+    bytes memory _replacedCalldata = swapHelper.getSwapCalldata(_token0, _token1, _newAmountIn, _newTo);
```
Comment (m100alpaca) [inline solidity/tests/swap-helper/SwapHelper.t.sol:73]: Do we expect this one?
``(address router, bytes memory data) = swapHelper.getSwapCalldata(_token0, _token1, _newAmountIn, _newTo);``

File: solidity/contracts/swap-helper/SwapHelper.sol [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  /// @param _destination The destination token
+  /// @param _amountIn The amount in
+  /// @param _to The destination address
+  /// @return _swapCalldata The modified swap calldata according to the input
+  function getSwapCalldata(
+    address _source,
+    address _destination,
+    uint256 _amountIn,
+    address _to
+  ) external view returns (bytes memory) {
```
Comment (m100alpaca) [inline solidity/contracts/swap-helper/SwapHelper.sol:23]: Should it be ``return (address router, bytes memory data)`` ?

**Review by m100alpaca — APPROVED** [review]

**Pull request merged**